omniledger/service/service_test.go

```diff
@@ -372,6 +372,75 @@ func TestService_StateChange(t *testing.T) {
 	require.Equal(t, latest, int64(n-1))
 }
 
+func TestService_DarcEvolutionFail(t *testing.T) {
+	s := newSer(t, 1, testInterval)
+	defer s.local.CloseAll()
+	defer closeQueues(s.local)
+
+	d2 := s.darc.Copy()
```

Comment:
This is a bit a surprising API, that you need to make a copy and then do "EvolveFrom" on the copy, referencing the source. Would it be possible/good if it worked like `d2 := s.darc.Evolve()` instead? So that Evolve() does the copy and the evolution at the same time?

(Sorry, I know this was already decided on before when darc was checked in.)

Reply:
I was also thinking about this issue. But doing `d2 := s.darc.Evolve(signers)` also won't work nicely because you can't change anything in the `s.darc`. You can do `d.Evolve()`, then change the darc, and then do `d.SignEvolution(signers)`. Which is also not so nice.

One solution I came up with, but haven't had time to implement yet, is to use the builder pattern. We use it a lot in java and I think it's quite elegant to configuring a complex object before finalising it.

Reply:
Created an issue - https://github.com/dedis/cothority/issues/1288